packages/lexical-code/src/__tests__/unit/LexicalCodeNodeTabs.test.ts

```diff
@@ -0,0 +1,222 @@
+import type {LexicalCommand, LineBreakNode, TabNode} from 'lexical';
+
+import {
+  $createCodeNode,
+  $isCodeNode,
+  registerCodeHighlighting,
+} from '@lexical/code';
+import {registerTabIndentation} from '@lexical/react/LexicalTabIndentationPlugin';
+import {registerRichText} from '@lexical/rich-text';
+import {
+  $createRangeSelection,
+  $getRoot,
+  $getSelection,
+  $isLineBreakNode,
+  $setSelection,
+  INDENT_CONTENT_COMMAND,
+  KEY_TAB_COMMAND,
+  OUTDENT_CONTENT_COMMAND,
+} from 'lexical';
+import {
+  initializeUnitTest,
+  tabKeyboardEvent,
+} from 'lexical/src/__tests__/utils';
+
+describe('LexicalCodeNode tests', () => {
+  initializeUnitTest((testEnv) => {
+    describe('Tabs', () => {
+      // Tests are described using the following convention
+      // > is the beginning of a line
+      // - is a Tab
+      // | is the selection anchor and focus (they may be collapsed)
+      // in the format [before, after, operation]
+      const suite = [
+        ['>|', '>-|', 'tab'],
+        ['>|', '>-|', 'indent'],
+        ['>|', '>|', 'outdent'],
+        ['>|line1', '>-|line1', 'tab'],
+        ['>|line1', '>-|line1', 'indent'],
+        ['>|line1', '>|line1', 'outdent'],
+        ['>-li|ne1', '>-li-|ne1', 'tab'],
+        ['>-li|ne1', '>--li|ne1', 'indent'],
+        ['>-li|ne1', '>li|ne1', 'outdent'],
+        ['>|-line1|', '>|--line1|', 'tab'],
+        ['>|-line1|', '>|--line1|', 'indent'],
+        ['>|-line1|', '>|line1|', 'outdent'],
+        ['>-li|ne1|', '>-li-|', 'tab'],
+        ['>-li|ne1|', '>--li|ne1|', 'indent'],
+        ['>-li|ne1|', '>li|ne1|', 'outdent'],
+        ['>|word1| word2', '>-| word2', 'tab'],
+        ['>|line1>|', '>|-line1>|', 'tab'],
+        ['>|line1>|', '>|-line1>|', 'indent'],
+        ['>|line1>|line2', '>|-line1>|line2', 'tab'],
+        ['>|line1>|line2', '>|-line1>|line2', 'indent'],
+        ['>line1|>line2|', '>-line1|>-line2|', 'tab'],
+        ['>line1|>line2|', '>-line1|>-line2|', 'indent'],
+        ['>|>li|ne1', '>|>-li|ne1', 'tab'],
+        ['>|>li|ne1', '>|>-li|ne1', 'indent'],
+        ['>|>|line1', '>|->|line1', 'tab'],
+        ['>|>|line1', '>|->|line1', 'indent'],
+        ['>|>>|line1', '>|>>|line1', 'tab'],
+        ['>|>>|line1', '>|>>|line1', 'indent'],
+        ['>|line1>li|ne2', '>|-line1>-li|ne2', 'tab'],
+        ['>|line1>li|ne2', '>|-line1>-li|ne2', 'indent'],
+        ['>|-line1>li|ne2', '>|line1>li|ne2', 'outdent'],
+        ['>li|ne1>li|ne2', '>-li|ne1>-li|ne2', 'tab'],
+        ['>li|ne1>li|ne2', '>-li|ne1>-li|ne2', 'indent'],
+        ['>-li|ne1>>-li|ne2', '>--li|ne1>>--li|ne2', 'tab'],
+        ['>-li|ne1>>-li|ne2', '>--li|ne1>>--li|ne2', 'indent'],
+        ['>-li|ne1>>-li|ne2', '>li|ne1>>li|ne2', 'outdent'],
+        ['>|-line1>->-line2|', '>|--line1>-->--line2|', 'tab'],
+      ];
+      suite.forEach((scenario) => {
+        test(`testing ${scenario[2]}: ${scenario[0]} => ${scenario[1]} `, async () => {
```

Comment:
This is probably how I would implement the directionality

```suggestion
      suite.forEach((scenario) => {
        ['forwards', 'backwards'].forEach((direction) => {
          test(`testing ${scenario[2]}: ${scenario[0]} => ${scenario[1]} ${direction}`, async () => {
```